Report apis

alxdarksage opened the pull request:
APIs for general reports (either at a study level or for individuals in a study).

alxdarksage committed:
Adding DAO and services for reports.

alxdarksage committed:
Report controller, service, and dao.

alxdarksage committed:
Revision after code review, and realizing some permissions and routing issues.

alxdarksage committed:
Cleanup.

alxdarksage committed:
Fix comment.

DwayneJengSage commented on app/org/sagebionetworks/bridge/dynamodb/DynamoReportData.java:
```diff
@@ -0,0 +1,54 @@
+    @DynamoDBRangeKey
+    @JsonSerialize(using = LocalDateToStringSerializer.class)
+    @Override
+    public LocalDate getDate() {
```
Do we have a unit test for serializing and deserializing DynamoReportData?

alxdarksage replied:
Adding.

DwayneJengSage commented on test/org/sagebionetworks/bridge/dynamodb/DynamoReportDataDaoTest.java:
```diff
@@ -0,0 +1,101 @@
+        dao.saveReportData(report1);
+        
+        DateRangeResourceList<? extends ReportData> results = dao.getReportData(
+                reportDataKey, START_DATE, END_DATE);
```
Is it worth adding a report outside of the date range so make sure the date range query works? What about a report with a different key?

alxdarksage replied:
Added both cases.

DwayneJengSage commented on app/org/sagebionetworks/bridge/dynamodb/DynamoReportDataDao.java:
```diff
@@ -0,0 +1,80 @@
+        DynamoDBScanExpression scan = new DynamoDBScanExpression()
+                .withFilterConditionEntry("key", new Condition()
+                        .withComparisonOperator(ComparisonOperator.EQ)
+                        .withAttributeValueList(new AttributeValue(key.getKeyString())));
```
Don't use a scan. Scan will read the entire table, regardless of which and how many entries match your filters. Instead, use a query.

alxdarksage replied:
I think at one point I was trying to delete all reports for all users at once, and gave up on this. Changing.

DwayneJengSage commented on app/org/sagebionetworks/bridge/models/DateRangeResourceList.java:
```diff
@@ -0,0 +1,31 @@
+
+import org.joda.time.LocalDate;
+
+public class DateRangeResourceList<T> {
```
Unit test to make sure this serializes into the format we expect?

alxdarksage replied:
Added serialization test. Not deserialization because we never deserialize it on the server, so didn't want to add @JsonCreator et al when we don't do it, to support the test.

alxdarksage committed:
Merge branch 'develop' of https://github.com/Sage-Bionetworks/BridgePF into report-apis

DwayneJengSage commented on app/org/sagebionetworks/bridge/services/ReportService.java:
```diff
@@ -0,0 +1,113 @@
+                .withReportType(reportType)
+                .withIdentifier(identifier)
+                .withStudyIdentifier(studyId).build();
+    }
```
This defeats the whole purpose of using the builder pattern. The point of the builder pattern is that when we add a new parameter, we don't have to update N^2 constructors and constructor usages. If we anticipate adding new keys (which I suspect is likely), we should remove this faux-constructor and just use the builder directly. If we are sure we'll never add new report keys, then we should get rid of the builder and just use the constructor directly.

alxdarksage replied:
I really don't care either way, removed.

DwayneJengSage commented on test/org/sagebionetworks/bridge/models/reports/ReportDataKeyTest.java:
```diff
@@ -0,0 +1,50 @@
+
+import com.fasterxml.jackson.databind.JsonNode;
+
+public class ReportDataKeyTest {
```
EqualsVerifier test?

alxdarksage replied:
Added

DwayneJengSage commented on test/org/sagebionetworks/bridge/play/controllers/ReportControllerTest.java:
```diff
@@ -0,0 +1,206 @@
+import play.test.Helpers;
+
+@RunWith(MockitoJUnitRunner.class)
+public class ReportControllerTest {
```
Is it worth adding tests where start and end date are null, to make sure our service doesn't crash in that scenario?

alxdarksage replied:
Yes, but in ReportServiceTest where dates are defaulted. I'll add null test here though too.

DwayneJengSage commented on app/org/sagebionetworks/bridge/play/controllers/ReportController.java:
```diff
@@ -0,0 +1,142 @@
+     * Reports for specific individuals accessible only to consented study participants.
+     */
+    public Result getParticipantReport(String identifier, String startDateString, String endDateString) {
+        UserSession session = getAuthenticatedAndConsentedSession();
```
Will we run into the same issue as mPower with multiple consents? Or do we plan to address that a different way?

alxdarksage replied:
Reports built against this service need to pass the correct headers in, and that's a client side issue. That's a bug in the existing mPower 1.3 app. Here we expect consented users and that's correct. They must be consented.

DwayneJengSage commented on test/org/sagebionetworks/bridge/services/ReportServiceTest.java:
```diff
@@ -0,0 +1,306 @@
+    @Test(expected = BadRequestException.class)
+    public void dateRangeTooWide() {
+        service.getParticipantReport(TEST_STUDY, IDENTIFIER, HEALTH_CODE, START_DATE, START_DATE.plusDays(46));
+    }
```
Should we also have the date validation tests and default date tests for Study reports as well?

alxdarksage replied:
Added.

alxdarksage committed:
Updating based on code review.

alxdarksage committed:
Adding an explicit check for headers to determine consent when retrieving a report.

DwayneJengSage commented on test/org/sagebionetworks/bridge/dynamodb/DynamoReportDataTest.java:
```diff
@@ -0,0 +1,50 @@
+        reportData.setData(objNode);
+        
+        String json = MAPPER.writeValueAsString(reportData);
+        ReportData deser = MAPPER.readValue(json, ReportData.class);
```
We want to validate the JSON to make sure there are no weird serialized JSON.

alxdarksage replied:
Okay, added that too.

DwayneJengSage commented on test/org/sagebionetworks/bridge/play/controllers/ReportControllerTest.java:
```diff
@@ -115,8 +132,16 @@ public void getParticipantReportDataNoDates() throws Exception {
+
+        controller.getParticipantReport("foo", START_DATE.toString(), END_DATE.toString());
+    }
+    
```
Test case to also verify language?

alxdarksage replied:
Added

alxdarksage committed:
- had to add a separate API for workers to add participant reports, because workers will not know the user ID of the account they are generating reports for;
- removed worker from the delete methods (I don't believe a worker will do this and in the participant case, it can't
- switched AccountDao for ParticipantService. If all you're doing is looking up a healthCode, it uses less resources to do this.

alxdarksage committed:
Added tests for ommitted or invalid language header.
